sockeye/train.py

```diff
@@ -132,278 +139,446 @@ def main():
     else:
         os.makedirs(output_folder)
 
+    return resume_training, training_state_dir
+
+
+def determine_context(args: argparse.Namespace, exit_stack: ExitStack) -> List[mx.Context]:
+    """
+    Determine the context we should run on (CPU or GPU).
+
+    :param args: Arguments as returned by argparse.
+    :param exit_stack: An ExitStack from contextlib.
+    :return: A list with the context(s) to run on.
+    """
+    if args.use_cpu:
+        logger.info("Device: CPU")
+        context = [mx.cpu()]
+    else:
+        num_gpus = utils.get_num_gpus()
+        check_condition(num_gpus >= 1,
+                        "No GPUs found, consider running on the CPU with --use-cpu "
+                        "(note: check depends on nvidia-smi and this could also mean that the nvidia-smi "
+                        "binary isn't on the path).")
+        if args.disable_device_locking:
+            context = utils.expand_requested_device_ids(args.device_ids)
```

Comment:
not a blocker, but if we're already factoring out the different parts into individual functions it would be nice to make the dependencies explicit, by passing in the required arguments instead of the entire namespace, e.g.:
```
def determine_context(use_cpu, disable_device_locking, device_ids, lock_dir, exit_stack: ExitStack) -> List[mx.Context]:
```
This way the individual functions could be used independent of argparse.
Anyway, we don't have to do this now and the current refactoring is also very valuable :)

Reply:
I'm ambivalent about this one. On the one hand it's good to remove the dependency of argparse. On the other hand, I find it convenient to pass the whole arguments and let the function pick what it needs. In this way we reduce the number of places that must be aware of the parameters. Think about the unpacking we would need to do for create_model_config.